File: services/backend/src/deskops_backend/domain/metrics_calculator.py
from deskops_backend.domain.models import TelemetryEvent


def calculate_summary(events: list[TelemetryEvent]) -> dict:
    total = len(events)
    success_count = sum(1 for e in events if e.result == "success")
    failed_count = sum(1 for e in events if e.result == "failed")
    review_count = sum(1 for e in events if e.result == "review")
    by_category: dict[str, int] = {}

    for event in events:
        by_category[event.category] = by_category.get(event.category, 0) + 1

    denominator = total if total > 0 else 1
    return {
        "total_events": total,
        "success_count": success_count,
        "failed_count": failed_count,
        "review_count": review_count,
        "success_rate": round(success_count / denominator, 4),
        "failure_rate": round(failed_count / denominator, 4),
        "review_rate": round(review_count / denominator, 4),
        "by_category": by_category,
    }

File: archive/legacy-node-backend/src/domain/metrics-calculator.mjs
export function calculateSummary(events) {
  const totals = {
    totalEvents: events.length,
    successCount: 0,
    failedCount: 0,
    reviewCount: 0,
    byCategory: {}
  };

  for (const event of events) {
    if (event.result === "success") totals.successCount += 1;
    if (event.result === "failed") totals.failedCount += 1;
    if (event.result === "review") totals.reviewCount += 1;
    totals.byCategory[event.category] = (totals.byCategory[event.category] ?? 0) + 1;
  }

  const denominator = totals.totalEvents || 1;
  return {
    ...totals,
    successRate: Number((totals.successCount / denominator).toFixed(4)),
    failureRate: Number((totals.failedCount / denominator).toFixed(4)),
    reviewRate: Number((totals.reviewCount / denominator).toFixed(4))
  };
}
